src/shears/pp/__init__.py

```diff
@@ -72,21 +75,132 @@ def cell_weights(
 
     If inplace is True, stores the resulting matrix in adata_sc.obsm[key_added]
     """
-    res = process_map(
-        _deconvolute,
-        (adata_bulk.layers[layer_bulk][i, :] for i in range(adata_bulk.shape[0])),
-        itertools.repeat(adata_sc.layers[layer_sc].T),
-        itertools.repeat(alpha_callback(adata_sc)),
-        itertools.repeat(random_state),
-        max_workers=n_jobs,
-        chunksize=10,
-        tqdm_class=tqdm,
-        total=adata_bulk.shape[0],
+
+    assert all(adata_sc.var_names == adata_bulk.var_names), "var_names are not in the same order or are not the same"
+
+    # convert sparse bulk layer to dense 1D arrays -> when obs > 1000 scanpy seems to auto convert to sparse ...
+    if sparse.issparse(adata_bulk.layers[layer_bulk]):
+        adata_bulk.layers[layer_bulk] = adata_bulk.layers[layer_bulk].toarray()
+
+    jobs = (
+        delayed(_deconvolute)(
+            adata_bulk.layers[layer_bulk][i, :],
+            adata_sc.layers[layer_sc].T,
+            alpha_callback(adata_sc),
+            random_state,
+        )
+        for i in range(adata_bulk.shape[0])
     )
 
-    res = pd.DataFrame(np.array(res).T, index=adata_sc.obs_names, columns=adata_bulk.obs_names)
+    weights_list = list(
+        _parallelize_with_joblib(
+            jobs, total=adata_bulk.shape[0], n_jobs=n_jobs, backend=backend
+        )
+    )
+
+    res = pd.DataFrame(
+        np.array(weights_list).T, index=adata_sc.obs_names, columns=adata_bulk.obs_names
+    )
 
     if inplace:
         adata_sc.obsm[key_added] = res
     else:
-        return res
+        return res_df
+
+
+def _block_iter(n_cells: int, n_bulk: int, block_size: int) -> Iterator[Tuple[int, int, int, int]]:
+    """
+    Compute the ranges of rows and columns for dividing an (n_cells × n_bulk) matrix into equally sized square chunks.
+    Each block is defined by (i0, i1, j0, j1).
+    """
+    for i0 in range(0, n_cells, block_size):
+        i1 = min(i0 + block_size, n_cells)
+        for j0 in range(0, n_bulk, block_size):
+            j1 = min(j0 + block_size, n_bulk)
+            yield i0, i1, j0, j1
+
+
+def _compute_block(sc_mat: np.ndarray, bulk_mat: np.ndarray, i0: int, i1: int, j0: int, j1: int) -> List[Tuple[float, float, int, int]]:
+    """
+    Calculate the Pearson correlation coefficient and its p-value for each (cell, bulk) pair in the given block.
+    Returns a list of tuples (r, p, global_cell_idx, global_bulk_idx).
+    """
+    res = []
+    for ii in range(i0, i1):
+        for jj in range(j0, j1):
+            r, p = pearsonr(sc_mat[ii], bulk_mat[jj])
+            res.append((r, p, ii, jj))
+    return res
+
```

Comment:
I don't really care that much, because this is only for comparison with the legacy scissor, so feel free to leave as is...

but shouldn't [numpy.corrcoef](numpy.corrcoef) directly compute the correlation between two arrays? I'd assume it to be more efficient and already use parallelism implicitly due to the underlying BLAS library.